### PR TITLE
Rational config handling

### PR DESCRIPTION
**Pull request by PDoakORNL**
To accomplish support for using pycicle with a personal fork and a PR destination branch other than master.  I tried to be careful to not break or change any current functionality.  For instance a good test would be hpx should be able to run with this as if it were master.

Unfortunately my existing CI instances weren't upgraded to the post merge master which did produce some minor breaking changes to the DCA CI on condo.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -159,6 +159,22 @@ $PYCICLE_ROOT/build
...
+An example is in config/dca_local
+#### It submits to a CDash server reverse tunneled to a localhost port on machine B (laptop)
+* still debugging this feature *
+Assuming cdash is setup on a httpd running on localhost:8080 of machine B.
+From machine B
+``` shell
+ssh -fN -R38080:localhost:8080 you@clusterA-login-node
+```
+If you have load balancing on log in nodes make sure to explicitly raise the reverse tunnel on the login node pycicle is running on.
+
```
Comment (biddisco) [inline README.md]: Doc improvements Wonderful

File: config/dca/cadesGPUCondo.cmake [hunk 2]
```diff
@@ -14,9 +13,9 @@ set(PYCICLE_MACHINE "local")
 set(PYCICLE_ROOT "/lustre/or-hydra/cades-cnms/epd/DCA_GPU_CI")
 # a flag that says if the machine can send http results to cdash
 set(PYCICLE_HTTP TRUE)
-# Method used to launch jobs "slurm", "pbs" or "direct" supported
-set(PYCICLE_JOB_LAUNCH "pbs")
-
+# Launch jobs using slurm rather than directly running them on the machine
+set(PYCICLE_SLURM FALSE)
+set(PYCICLE_PBS TRUE)
```
Comment (biddisco) [inline config/dca/cadesGPUCondo.cmake]: I think the `set(PYCICLE_JOB_LAUNCH "pbs")` approach is nicer than having a bool, for each possiblity. This was changed in my last big update and I suspect you started adding these tweaks before I merged your other stuff.

File: config/dca/cadesGPUCondo.cmake [hunk 3]
```diff
@@ -110,7 +112,7 @@ string(CONCAT CTEST_BUILD_OPTIONS ${CTEST_BUILD_OPTIONS}
 # Setup a slurm job submission template
 # note that this is intentionally multiline
 #######################################################################
-set(PYCICLE_JOB_SCRIPT_TEMPLATE "#!/bin/bash
+set(PYCICLE_PBS_TEMPLATE "#!/bin/bash
```
Comment (biddisco) [inline config/dca/cadesGPUCondo.cmake]: same here - I renamed it to PYCICLE_JOB_SCRIPT_TEMPLATE to make it easier to search/replace and switch between slurm/pbs /other

File: config/dca/cadesGPUCondo.cmake [hunk 4]
```diff
@@ -1,5 +1,4 @@
-#  Copyright (c) 2018      Peter Doak
-#  Copyright (c) 2017-2018 John Biddiscombe
+#  Copyright (c) 2018 John Biddiscombe, Peter Doak
```
Comment (biddisco) [inline config/dca/cadesGPUCondo.cmake]: trivial detail, but I quite like the hpx convention where each contributor puts his name on a separate line and edits the dates accordingly.

File: dashboard_script.cmake [hunk 5]
```diff
@@ -92,15 +94,17 @@ endif()
 #####################################################################
 # if this is a PR to be merged with master for testing
 #####################################################################
-if (NOT PYCICLE_PR STREQUAL "master")
+if (NOT PYCICLE_PR STREQUAL "${PYCICLE_MASTER}")
```
Comment (biddisco) [inline dashboard_script.cmake]: good catch

File: pycicle_params.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            'PYCICLE_MACHINE',
+            'PYCICLE_HTTP',
+            'PYCICLE_JOB_LAUNCH',
+            'PYCICLE_GITHUB_MASTER_BRANCH',
+            'PYCICLE_CDASH_SERVER_NAME',
+            'PYCICLE_CDASH_PROJECT_NAME',
+            'PYCICLE_CDASH_HTTP_PATH',
+            'PYCICLE_BUILD_STAMP',
+            'PYCICLE_COMPILER_SETUP',
+            'PYCICLE_SLURM',
```
Comment (biddisco) [inline pycicle_params.py]: same comment here - see refactoroed slurm/pbs option

File: pycicle_params.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
+#  Copyright (c) 2017-2018 John Biddiscombe, Peter Doak
```
Comment (biddisco) [inline pycicle_params.py]: wow. I'd better test this right away. all the settings in one place. good idea

File: config/dca/cadesGPUCondo.cmake (hunk 2 above)
Comment (PDoakORNL) [reply config/dca/cadesGPUCondo.cmake]: I didn't test the dca master branch CI so I missed this.

File: config/dca/cadesGPUCondo.cmake (hunk 4 above)
Comment (PDoakORNL) [reply config/dca/cadesGPUCondo.cmake]: Yest that's more informative, I'll switch it back

File: pycicle_params.py (hunk 6 above)
Comment (PDoakORNL) [reply pycicle_params.py]: Yes part of the point here is to support some config error catching.

File: config/dca/cadesGPUCondo.cmake (hunk 4 above)
Comment (PDoakORNL) [reply config/dca/cadesGPUCondo.cmake]: So last person to update file on top?

**Comment by PDoakORNL** [comment]
Ok fixed those issues.

**Comment by biddisco** [comment]
```
jb-s76:~/pycicle$ python3 ~/src/pycicle/pycicle.py -m jb-laptop -P hpx -d
Traceback (most recent call last):
  File "/home/biddisco/src/pycicle/pycicle.py", line 26, in <module>
    from pycicle_params import PycicleParams
  File "/home/biddisco/src/pycicle/pycicle_params.py", line 74
    m = re.findall(setting + ur"\s*\"(.+?)\"", line)
                                            ^
SyntaxError: invalid syntax
```
must be something to do with python 2 vs python 3. I use python3 on my machines.

I'll try to fix it ...

**Comment by biddisco** [comment]
so it seems python 3 allows u"string" or r"string", but not ur"string", so I'll use u"string"

I've got a problem with default args. pycicle compiler_type is set to None and the default "gcc" is not substituted. Might as well make that a compulsory setting and remove the default value. Only needs me to add gcc to my jb-laptop config.

**Pull request merged**